File: 백트래킹/사다리조작.java
import java.util.ArrayList;
import java.util.Scanner;

public class 사다리조작 {
    static int N, M, H, CNT;
    static int[][] arr;
    static ArrayList<int[]> pos;
    static int ans;
    static int cnt;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        N = sc.nextInt(); // 세로선의 개수
        M = sc.nextInt(); // 주어진 가로선의 개수
        H = sc.nextInt(); // 가로선을 놓을 수 있는 높이의 개수
        
        arr = new int[H + 1][N + 2]; // 배열 초기화 (좌우에 공백 추가)
        
        // [1] 미리 놓아진 가로선 입력
        for (int i = 0; i < M; i++) {
            int ti = sc.nextInt();
            int tj = sc.nextInt();
            arr[ti][tj] = 1; // 가로선이 있는 위치 표시
        }
        
        // [2] 가로선을 추가할 후보 위치 저장
        pos = new ArrayList<>();
        for (int i = 1; i <= H; i++) {
            for (int j = 1; j <= N; j++) {
                if (arr[i][j] == 0 && arr[i][j - 1] == 0 && arr[i][j + 1] == 0) {
                    pos.add(new int[]{i, j});
                }
            }
        }
        
        CNT = pos.size();
        
        // [3] 추가할 가로선 개수를 0부터 3까지 실행, 불가능하면 -1 출력
        for (cnt = 0; cnt <= 3; cnt++) {
            ans = 0;
            backtrack(0, 0);
            if (ans == 1) {
                System.out.println(cnt);
                return;
            }
        }
        
        System.out.println(-1);
    }

    // 모든 세로선이 시작 위치와 동일한 위치에 도착하는지 확인하는 함수
    static int check() {
        for (int sj = 1; sj <= N; sj++) {
            int j = sj;
            for (int i = 1; i <= H; i++) {
                if (arr[i][j] == 1) { // 오른쪽으로 이동
                    j++;
                } else if (j > 1 && arr[i][j - 1] == 1) { // 왼쪽으로 이동
                    j--;
                }
            }
            if (j != sj) {
                return 0; // 하나라도 다르면 실패
            }
        }
        return 1;
    }

    // 가로선을 추가하면서 모든 가능한 경우를 백트래킹으로 탐색
    static void backtrack(int n, int s) {
        if (ans == 1) {
            return; // 이미 답을 찾으면 더 이상 진행하지 않음
        }

        if (n == cnt) { // 필요한 개수의 가로선을 모두 선택했을 때
            if (check() == 1) {
                ans = 1;
            }
            return;
        }

        for (int j = s; j < CNT; j++) {
            int[] position = pos.get(j);
            int ti = position[0];
            int tj = position[1];
            
            // 왼쪽과 오른쪽에 가로선이 없는 경우만 가로선 추가 가능
            if (arr[ti][tj - 1] == 0 && arr[ti][tj + 1] == 0) {
                arr[ti][tj] = 1; // 가로선 추가
                backtrack(n + 1, j + 1); // 다음 가로선 위치로 진행
                arr[ti][tj] = 0; // 가로선 제거 (백트래킹)
            }
        }
    }
}

File: 백트래킹/사다리조작.py
def check():
    for sj in range(1, N+1):    # 모든 sj값이 끝날때도 같아야 성공!
        j = sj
        for i in range(1, H+1):
            if arr[i][j] == 1:    # 오른쪽으로 이동
                j+=1
            elif arr[i][j-1] == 1:# 왼쪽으로 이동
                j-=1
        if j != sj:
            return 0            # 하나라도 다르다면 실패
    return 1

def backtrack(n, s):
    global ans
    if ans == 1:      # 가지치기
        return

    if n == cnt:      # 모든 개수를 선택완료
        if check() == 1:
            ans = 1
        return

    for j in range(s, CNT):
        ti,tj = pos[j]
        # 왼쪽도 사다리가 아니고, 오른쪽도 사다리가 아니면 가능!
        if arr[ti][tj - 1] == 0 and arr[ti][tj + 1] == 0:
            arr[ti][tj] = 1
            backtrack(n + 1, j + 1)
            arr[ti][tj] = 0

N, M, H = map(int, input().split())

# [1] 사다리 입력 받기 (문제 좌표와 직관적으로 일치)
arr = [[0] * (N+2) for _ in range(H+1)]   # 좌우 공백추가(범위체크X)
for _ in range(M):
    ti,tj = map(int, input().split())   # 미리 놓아져 있는 사다리표시
    arr[ti][tj] = 1

# [2] 사다리 놓을 후보 저장
pos = []
for i in range(1, H+1):
    for j in range(1, N+1):
        if arr[i][j]==0 and arr[i][j-1]==0 and arr[i][j+1]==0:
            pos.append((i,j))
CNT = len(pos)

# [3] 추가하는 사다리개수 0~3 실행 => 안되면 -1이 정답
for cnt in range(4):
    ans = 0
    backtrack(0,0)    # CNT개에서 cnt개수를 뽑는 모든 조합
    if ans == 1:
        ans = cnt
        break
else:           # 발견하지 못했다면 break안했음..
    ans = -1
print(ans)